lib/rex/post/mysql/ui/console/command_dispatcher/client.rb

```diff
@@ -17,6 +17,23 @@ def name
             'MySQL Client'
           end
 
+          # @param [Object] query
+          # @return [Hash{Symbol->Symbol | Hash{Symbol->Array<Exception>}}]
+          def run_query(query)
+            begin
+              result = client.query(query)
+            rescue Mysql::ClientError::ServerLost => e
+              elog("Running query '#{query}' failed on session #{self.inspect}", error: e)
+              session.alive = false
+              return { status: :error, result: { errors: [e] } }
+            rescue ::RuntimeError, ::StandardError => e
+              elog("Running query '#{query}' failed on session #{self.inspect}", error: e)
+              return { status: :error, result: { errors: [e] } }
+            end
+
+            super
```

Comment:
It looks like you're calling off to query twice here, which is wrong 👀 

Maybe just call `super` - then conditionally kill the session based on the `errors` response

Reply:
Looks like this also breaks the assumptions of the EOF handling for postgres etc

Reply:
Maybe something like this:

```
diff --git a/lib/rex/post/mysql/ui/console/command_dispatcher/client.rb b/lib/rex/post/mysql/ui/console/command_dispatcher/client.rb
index 81cbd0b203..5e42f3ecf0 100644
--- a/lib/rex/post/mysql/ui/console/command_dispatcher/client.rb
+++ b/lib/rex/post/mysql/ui/console/command_dispatcher/client.rb
@@ -39,6 +39,14 @@ module Rex
             # meaning we don't have that in the Result object.
             { rows: result.entries, columns: result.fields.each.map(&:name), errors: [] }
           end
+
+          def handle_error(e)
+            case e
+            when Mysql::ClientError::ServerLost
+              _close_session
+            end
+            super
+          end
         end
       end
     end
diff --git a/lib/rex/post/sql/ui/console/command_dispatcher/client.rb b/lib/rex/post/sql/ui/console/command_dispatcher/client.rb
index 5fffcaa26f..fa776db67e 100644
--- a/lib/rex/post/sql/ui/console/command_dispatcher/client.rb
+++ b/lib/rex/post/sql/ui/console/command_dispatcher/client.rb
@@ -120,7 +120,7 @@ module Rex
               def run_query(query)
                 begin
                   result = client.query(query)
-                rescue ::RuntimeError, ::StandardError => e
+                rescue => e
                   elog("Running query '#{query}' failed on session #{self.inspect}", error: e)
                   return { status: :error, result: { errors: [e] } }
                 end
@@ -177,6 +177,9 @@ module Rex
                   print_line(formatted_result.to_s)
                 when :error
                   print_error "Query has failed. Reasons: #{result[:result][:errors].join(', ')}"
+                  result[:result][:errors].each do |error|
+                    handle_error(error)
+                  end
                 else
                   elog "Unknown query status: #{result[:status]}"
                   print_error "Unknown query status: #{result[:status]}"
@@ -188,6 +191,20 @@ module Rex
 
                 query.lines.each.map { |line| line.chomp.chomp('\\').strip }.reject(&:empty?).compact.join(' ')
               end
+
+              def handle_error(e)
+                case e
+                when EOFError
+                  _close_session
+                end
+                end
+              end
+
+              # TODO: This might already exist
+              def _close_session
+                session.alive = false
+                session.client.interacting = false if session.client && session.client.respond_to?(:interacting)
+              end
             end
           end
         end

```

Reply:
Got those changes added. Thanks for taking a look 👍